### PR TITLE
Add MirroredStrategy test for TF2 native profiling

### PR DESCRIPTION
**Pull request by ndodda-amazon**
### Description of changes:
Add MirroredStrategy distributed training test for TF2 native profiling.
- Runs end to end distributed training with MirroredStrategy in addition to the existing configurations tested for TF2 native profiling
- Fix bug to support tensor collection with MirroredStrategy and TF2 native training
- Fix bug in Pyinstrument profiling to support distributed training

#### Style and formatting:

I have run `pre-commit install` to ensure that auto-formatting happens with every commit.

#### Issue number, if available

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: smdebug/profiler/python_profiler.py [hunk 1]
```diff
@@ -225,6 +227,8 @@ class PyinstrumentPythonProfiler(PythonProfiler):
     def _enable_profiler(self):
         """Enable the pyinstrument profiler.
         """
+        if sys.getprofile() is not None:
```
Comment (NihalHarish) [inline smdebug/profiler/python_profiler.py:230]: Why is this change required to support mirrored strategy?
Add a comment if this is explicitly required for distributed training.

File: tests/core/json_configs/test_native_tf2_cprofile_profiler_config_parser.json [hunk 2]
```diff
@@ -2,7 +2,7 @@
     "ProfilingParameters": {
         "ProfilerEnabled": true,
         "LocalPath": "/tmp/test",
-        "DetailedProfilingConfig": "{\"StartStep\": 2, \"NumSteps\": 2}",
-        "PythonProfilingConfig": "{\"StartStep\": 5, \"NumSteps\": 2, \"ProfilerName\": \"cProfile\"}"
+        "DetailedProfilingConfig": "{\"StartStep\": 0, \"NumSteps\": 2}",
+        "PythonProfilingConfig": "{\"StartStep\": 0, \"NumSteps\": 2, \"ProfilerName\": \"cProfile\"}"
```
Comment (NihalHarish) [inline tests/core/json_configs/test_native_tf2_cprofile_profiler_config_parser.json:6]: Why?

File: tests/core/json_configs/test_native_tf2_pyinstrument_profiler_config_parser.json [hunk 3]
```diff
@@ -2,7 +2,7 @@
     "ProfilingParameters": {
         "ProfilerEnabled": true,
         "LocalPath": "/tmp/test",
-        "DetailedProfilingConfig": "{\"StartStep\": 2, \"NumSteps\": 2}",
-        "PythonProfilingConfig": "{\"StartStep\": 5, \"NumSteps\": 2, \"ProfilerName\": \"Pyinstrument\"}"
+        "DetailedProfilingConfig": "{\"StartStep\": 0, \"NumSteps\": 2}",
+        "PythonProfilingConfig": "{\"StartStep\": 0, \"NumSteps\": 2, \"ProfilerName\": \"Pyinstrument\"}"
```
Comment (NihalHarish) [inline tests/core/json_configs/test_native_tf2_pyinstrument_profiler_config_parser.json:6]: Same question as above.
Why do we need to change the json config?

File: tests/fixtures/tf2_training.py [hunk 4]
```diff
@@ -92,6 +89,22 @@ def call(self, x, training=None):
...
+        ModelType.FUNCTIONAL: _get_tf2_mnist_functional_model,
+        ModelType.SUBCLASSED: _get_tf2_mnist_subclassed_model,
+    }
+
+    def _get_model_and_optimizer(model_type):
+        return model_dict[model_type]()
+
+    return _get_model_and_optimizer
+
+
```
Comment (NihalHarish) [inline tests/fixtures/tf2_training.py:107]: Thanks!

File: tests/profiler/tensorflow2/test_native_tf2_profiler.py [hunk 5]
```diff
@@ -31,65 +34,52 @@
...
 
 @pytest.fixture
-def native_tf2_cprofile_profiler_config_parser(config_folder, monkeypatch):
-    config_path = os.path.join(
-        config_folder, "test_native_tf2_cprofile_profiler_config_parser.json"
-    )
-    monkeypatch.setenv("SMPROFILER_CONFIG_PATH", config_path)
-    return ProfilerConfigParser(FRAMEWORK.TENSORFLOW)
+def native_tf2_cprofile_profiler_config_path(config_folder, monkeypatch):
+    return os.path.join(config_folder, "test_native_tf2_cprofile_profiler_config_parser.json")
```
Comment (NihalHarish) [inline tests/profiler/tensorflow2/test_native_tf2_profiler.py:38]: Why do we not need to set the env variable anymore?

**Review by NihalHarish — CHANGES_REQUESTED** [review]

File: tests/profiler/tensorflow2/test_native_tf2_profiler.py [hunk 6]
```diff
@@ -31,65 +34,52 @@
...
 
 @pytest.fixture
-def native_tf2_pyinstrument_profiler_config_parser(config_folder, monkeypatch):
-    config_path = os.path.join(
-        config_folder, "test_native_tf2_pyinstrument_profiler_config_parser.json"
-    )
+def native_tf2_pyinstrument_profiler_config_path(config_folder, monkeypatch):
+    return os.path.join(config_folder, "test_native_tf2_pyinstrument_profiler_config_parser.json")
     monkeypatch.setenv("SMPROFILER_CONFIG_PATH", config_path)
     return ProfilerConfigParser(FRAMEWORK.TENSORFLOW)
```
Comment (NihalHarish) [inline tests/profiler/tensorflow2/test_native_tf2_profiler.py:46]: Are there two return statements in this function?

**Review by NihalHarish — CHANGES_REQUESTED** [review]

File: smdebug/profiler/python_profiler.py (hunk 1 above)
Comment (ndodda-amazon) [reply smdebug/profiler/python_profiler.py:230]: See PR description.

> Fix bug in Pyinstrument profiling to support distributed training

File: tests/core/json_configs/test_native_tf2_cprofile_profiler_config_parser.json (hunk 2 above)
Comment (ndodda-amazon) [reply tests/core/json_configs/test_native_tf2_cprofile_profiler_config_parser.json]: Thanks for catching, forgot to revert these changes.

File: tests/core/json_configs/test_native_tf2_pyinstrument_profiler_config_parser.json (hunk 3 above)
Comment (ndodda-amazon) [reply tests/core/json_configs/test_native_tf2_pyinstrument_profiler_config_parser.json]: Thanks for catching, forgot to revert these changes.

File: tests/profiler/tensorflow2/test_native_tf2_profiler.py (hunk 5 above)
Comment (ndodda-amazon) [reply tests/profiler/tensorflow2/test_native_tf2_profiler.py]: Thanks for catching, forgot to revert these changes.

File: tests/profiler/tensorflow2/test_native_tf2_profiler.py (hunk 6 above)
Comment (ndodda-amazon) [reply tests/profiler/tensorflow2/test_native_tf2_profiler.py]: Thanks for catching, forgot to revert these changes.

File: tests/profiler/tensorflow2/test_native_tf2_profiler.py [hunk 7]
```diff
@@ -122,34 +117,35 @@ def _verify_timeline_files(out_dir):
...
+            TRACE_DIRECTORY_FORMAT
+        )
 
-    with open(file) as timeline_file:
-        events_dict = json.load(timeline_file)
+        with open(file) as timeline_file:
+            events_dict = json.load(timeline_file)
 
-    assert events_dict is not None
+        assert events_dict is not None
```
Comment (NihalHarish) [inline tests/profiler/tensorflow2/test_native_tf2_profiler.py:135]: Validate the content of the files. Atleast the keys to start with?

File: smdebug/profiler/python_profiler.py (hunk 1 above)
Comment (NihalHarish) [reply smdebug/profiler/python_profiler.py:230]: Add comments explaining the underlying logic.

**Review by NihalHarish — CHANGES_REQUESTED** [review]

File: smdebug/profiler/python_profiler.py (hunk 1 above)
Comment (ndodda-amazon) [reply smdebug/profiler/python_profiler.py]: Decided to dive deeper and realized that it doesn't make sense for some python profiling functionality (profiling during the step) to be distributed across the workers and for some python profiling functionality (pre-step-zero and post-hook-close) to be called before and after training. All python profiling calls need to be distributed across all of the workers. Once I made that change, I no longer saw this edge case occurring and removed this condition.

TL;DR wrote an incorrect use case earlier, this edge case is no longer valid.

File: tests/profiler/tensorflow2/test_native_tf2_profiler.py (hunk 7 above)
Comment (ndodda-amazon) [reply tests/profiler/tensorflow2/test_native_tf2_profiler.py]: Done.

File: smdebug/profiler/python_profiler.py (hunk 1 above)
Comment (NihalHarish) [reply smdebug/profiler/python_profiler.py]: Remove this code snippet then

File: tests/profiler/tensorflow2/test_native_tf2_profiler.py [hunk 8]
```diff
@@ -163,38 +171,54 @@ def test_native_tf2_profiling(
...
+
+    optimizer = hook.wrap_optimizer(optimizer)
+    _training_loop(
+        hook, profiler_config_parser, model, optimizer, mnist_dataset, train_step_func, strategy
+    )
 
-    # Sanity check debugger output
-    _verify_tensor_names(out_dir)
+    # Sanity check debugger output.
+    # _verify_tensor_names(out_dir)
```
Comment (NihalHarish) [inline tests/profiler/tensorflow2/test_native_tf2_profiler.py:208]: Did you mean to uncomment this or remove it?

**Review by NihalHarish — APPROVED** [review]

**Review by NihalHarish — APPROVED** [review]
